src/stub/preprocessor/forward_declarations.rs

```diff
@@ -0,0 +1,120 @@
+use super::Renderable;
+use crate::stub::{Cmd, Stub, VarType, VariableCommand};
+
+const ALPHABET: [char; 18] = [
+    'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
+];
+
+/// Change the Stub structure into: [ReadDeclarations, MainContents(old_cmds)]
+/// This is relevant for Pascal.
+#[derive(Debug, Clone)]
+struct MainWrapper {
+    // Read declarations that should go on top of the main function.
+    // render declaration: int c;
+    // render read (usual): int c;\nscanf("%d", c);
+    pub forward_declarations: Vec<VariableCommand>,
+    // The main function contents.
+    pub main_content: Vec<Cmd>,
+}
+
+/// Edit stub to allow for Pascal-style forward declarations.
+///
+/// Wraps all of the commands in a stub that contains:
+/// - Forward declarations
+/// - The rest of the code
+///
+/// Traverses through the stub commands, taking all declared variables.
+/// Leaves only one MainWrapper command in the stub.
+/// Introduces two new templates:
+/// - `forward_declaration` - similar to a read_one declares one single
+///   variable, includes all the fields in a VariableCommand (not nested under
+///   `var`)
+/// - `main_wrapper` - wraps all of the code, contains `forward_declarations`
+///   (the above resource, rendered) and `main_content`
+pub fn transform(stub: &mut Stub) {
+    let mut max_nested_depth = 0;
+
+    let mut forward_declarations: Vec<VariableCommand> = stub
+        .commands
+        .iter()
+        .filter_map(|cmd| {
+            let (cmd, nested_depth) = unpack_cmd(cmd, 0);
+
+            if nested_depth > max_nested_depth {
+                max_nested_depth = nested_depth;
+            }
+
+            if let Cmd::Read(var_cmds)
+            | Cmd::LoopLine {
+                variables: var_cmds, ..
+            } = cmd
+            {
+                Some(var_cmds.into_iter())
+            } else {
+                None
+            }
+        })
+        .flatten()
+        .collect();
+
+    // Add also the forloop indices to the declarations.
+    forward_declarations.extend(ALPHABET[0..max_nested_depth].iter().map(|loop_var| VariableCommand {
+        ident: loop_var.to_string(),
+        var_type: VarType::Int,
+        max_length: None,
+        input_comment: String::new(),
+    }));
+
+    let mut unique_forward_declarations: Vec<VariableCommand> = Vec::new();
+    let mut seen_idents: Vec<String> = Vec::new();
+
+    for var_cmd in &forward_declarations {
+        if !seen_idents.contains(&var_cmd.ident) {
+            seen_idents.push(var_cmd.ident.clone());
+            unique_forward_declarations.push(var_cmd.clone());
+        }
+    }
+
+    let wrapper = MainWrapper {
+        forward_declarations: unique_forward_declarations,
```

Comment:
If I understand this right it's just removing forward declarations that have the same identifier as another forward declaration? That could be significantly simplified by using the [`Vec::retain`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain) method. This also avoids some of the unnecessary String allocation.

```suggestion
    let mut seen = std::collections::BTreeSet::<String>::new();
    forward_declarations.retain(|var_cmd| seen.insert(var_cmd.ident.clone()));

    let wrapper = MainWrapper {
        forward_declarations,
```

There's probably a way to make `seen` only store references too but I'm too lazy to figure it out and it doesn't *really* matter with only a handful of variables in a stub.